src/r1cs_nark_pcd/help_circuit.rs

```diff
@@ -0,0 +1,273 @@
+    ASForR1CSNarkVerifierGadget, AccumulatorInstanceVar, InputInstanceVar,
+};
+use ark_accumulation::r1cs_nark_as::{AccumulatorInstance, InputInstance};
+use ark_ec::CurveCycle;
+use ark_ff::{PrimeField, Zero};
+use ark_r1cs_std::alloc::AllocVar;
+use ark_r1cs_std::bits::boolean::Boolean;
+use ark_r1cs_std::eq::EqGadget;
+use ark_r1cs_std::fields::fp::FpVar;
+use ark_relations::ns;
+use ark_relations::r1cs::{
+    ConstraintSynthesizer, ConstraintSystem, ConstraintSystemRef, SynthesisError,
+};
+use ark_sponge::constraints::CryptographicSpongeVar;
+use ark_sponge::{absorb, absorb_gadget, Absorbable, CryptographicSponge};
+use ark_std::marker::PhantomData;
+
+/// A circuit used to verify that the accumulation of arguments about the main circuit was computed
+/// correctly.
+#[derive(Derivative)]
+#[derivative(Clone(bound = "E: CurveCycle"))]
+pub(crate) struct HelpCircuit<E, PC, P>
+where
+    E: CurveCycle,
+    MainField<E>: PrimeField + Absorbable<MainField<E>>,
+    HelpField<E>: PrimeField + Absorbable<HelpField<E>>,
+    MainAffine<E>: Absorbable<HelpField<E>>,
+    HelpAffine<E>: Absorbable<MainField<E>>,
+    PC: R1CSNarkPCDConfig<E>,
+    P: PCDPredicate<MainField<E>>,
+{
+    /// The key for verifying the accumulation of arguments about the main circuit.
+    pub(crate) main_avk: r1cs_nark_as::VerifierKey,
+
+    /// The accumulation input instances of the arguments about the main circuit in the previous
+    /// step.
+    pub(crate) main_accumulation_input_instances: Option<Vec<InputInstance<MainAffine<E>>>>,
+
+    /// The old accumulators that have accumulated arguments about the main circuit in the previous
+    /// step.
+    pub(crate) main_old_accumulator_instances: Option<Vec<AccumulatorInstance<MainAffine<E>>>>,
+
+    /// The new accumulator computed from accumulating arguments and accumulators of the main
+    /// circuit of the previous step.
+    pub(crate) main_new_accumulator_instance: AccumulatorInstance<MainAffine<E>>,
+
+    /// The new proof computed from accumulating arguments and accumulators of the main
+    /// circuit of the previous step.
+    pub(crate) main_accumulation_proof: r1cs_nark_as::Proof<MainAffine<E>>,
+
+    #[doc(hidden)]
+    pub(crate) _config_phantom: PhantomData<PC>,
+
+    #[doc(hidden)]
+    pub(crate) _predicate_phantom: PhantomData<P>,
+}
+
+impl<E, PC, P> HelpCircuit<E, PC, P>
+where
+    E: CurveCycle,
+    MainField<E>: PrimeField + Absorbable<MainField<E>>,
+    HelpField<E>: PrimeField + Absorbable<HelpField<E>>,
+    MainAffine<E>: Absorbable<HelpField<E>>,
+    HelpAffine<E>: Absorbable<MainField<E>>,
+    PC: R1CSNarkPCDConfig<E>,
+    P: PCDPredicate<MainField<E>>,
+{
+    /// Returns the public input size of the help circuit.
+    pub(crate) fn public_input_size() -> usize {
+        let cs = ConstraintSystem::<HelpField<E>>::new_ref();
+        let _hash = FpVar::new_input(cs.clone(), || Ok(HelpField::<E>::zero())).unwrap();
+        return cs.num_instance_variables();
```

Comment:
If we know we're only ever allocating one variable, we can hard code the result of `cs.num_instance_variables()`. If that can change, we should instead invoke the appropriate code (eg: `SpongeOutout::new_input(cs.clone(), || Ok(dummy_output))`).

Reply:
I agree that we should invoke the appropriate code rather than directly using the HelpField. 

However, wouldn't it be a poor choice to break the abstraction of the ConstraintSystem by hard coding the expected number of instance variables? If I am remembering correctly, one implementation detail of ConstraintSystem is that it adds a one when initializing the instance variables. However, such detail isn't very relevant to this library (we only care that we are using the outputs of CS) and if that implementation detail ever changes (would it?), then it may introduce some difficult to find bugs. 

Considering this point, would we still want to hard code the number of instance variables?